File: src/Feb2021/sorting/java/QuickSort.java
package Feb2021.sorting.java;

import java.util.Scanner;

public class QuickSort {

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int numberOfCases = sc.nextInt();

        while (numberOfCases-- > 0) {
            int n = sc.nextInt();
            int[] arr = new int[n];
            for (int i = 0; i < n; i++) {
                arr[i] = sc.nextInt();
            }

            performQuickSort(arr, 0, n-1);
        }
    }

    private static void performQuickSort(int[] arr, int low, int high) {
        if (low < high) {
            int pivot = partition(arr, low, high);

            performQuickSort(arr, low, pivot-1);
            performQuickSort(arr, pivot+1, high);
        }
    }

    private static int partition(int[] arr, int low, int high) {

        return low;
    }
}

File: src/Feb2021/sorting/kotlin/QuickSort.kt
package Feb2021.sorting.kotlin

fun main(args: Array<String>) {
    val input = readLine()
    val inputList: ArrayList<Int> =
            ArrayList(input!!.split(" ").map { it.toInt() })

    quickSort(inputList.toIntArray(),
                0, inputList.size-1)
    printSortedArray(inputList)
}

fun printSortedArray(inputList: java.util.ArrayList<Int>) {
    for (i in 0 until inputList.size) {
        print("${inputList[i]} ")
    }
    println()
}

fun quickSort(toArray: IntArray, low: Int, high: Int) {

    if (low < high) {
        val pivot = partition(toArray, low, high)

        quickSort(toArray, low, pivot-1)
        quickSort(toArray, pivot+1, high)
    }
}

fun partition(toArray: IntArray, low: Int, high: Int): Int {
    val pivot = toArray[high]
    var i: Int = low - 1
    for (j in low until high) {
        if(toArray[j] < pivot) {
            i++
            val temp = toArray[i]
            toArray[i] = toArray[j]
            toArray[j] = temp
        }
    }

    val temp = toArray[i+1]
    toArray[i+1] = toArray[high]    // high or pivot
    toArray[high] = temp

    return i+1
}

